File: server/models/product_image.ts
'use strict';

import { Model, DataTypes, Sequelize } from 'sequelize';

interface ProductImageAttributes {
  id: string;
  product_id: string;
  image_url: string;
  is_primary: boolean;
  display_order: number;
  created_at: Date;
  updated_at: Date;
}

export class ProductImage extends Model<ProductImageAttributes> implements ProductImageAttributes {
  public id!: string;
  public product_id!: string;
  public image_url!: string;
  public is_primary!: boolean;
  public display_order!: number;
  public created_at!: Date;
  public updated_at!: Date;

  // Static methods
  static associate(models: any) {
    ProductImage.belongsTo(models.Product, {
      foreignKey: 'product_id',
      as: 'product',
    });
  }
}

export function initProductImage(sequelize: Sequelize): void {
  ProductImage.init(
    {
      id: {
        type: DataTypes.STRING(36),
        defaultValue: DataTypes.UUIDV4,
        primaryKey: true,
        field: 'id'
      },
      product_id: {
        type: DataTypes.STRING(36),
        allowNull: false,
        references: {
          model: 'products',
          key: 'id'
        },
        field: 'product_id'
      },
      image_url: {
        type: DataTypes.STRING,
        allowNull: false,
        field: 'image_url'
      },
      is_primary: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
        field: 'is_primary'
      },
      display_order: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
        field: 'display_order'
      },
      created_at: {
        type: DataTypes.DATE,
        field: 'created_at'
      },
      updated_at: {
        type: DataTypes.DATE,
        field: 'updated_at'
      }
    },
    {
      sequelize,
      modelName: 'ProductImage',
      tableName: 'product_images',
      timestamps: true,
      underscored: false, // Don't use automatic conversion
      
      // Define timestamps explicitly
      createdAt: 'created_at',
      updatedAt: 'updated_at',
      
      // Explicitly define field mappings to avoid camelCase/snake_case confusion
      indexes: [
        {
          fields: ['product_id']
        }
      ]
    }
  );
}

File: server/models/product_image.py
import uuid
from sqlalchemy.sql import func
from sqlalchemy import SmallInteger
from . import db

class ProductImage(db.Model):
    __tablename__ = 'product_images'
    
    id = db.Column(db.String(36), primary_key=True, default=lambda: str(uuid.uuid4()))
    product_id = db.Column(db.String(36), db.ForeignKey('products.id', ondelete='CASCADE'), nullable=False)
    image_url = db.Column(db.String(255), nullable=False)
    is_primary = db.Column(SmallInteger, default=0) # TINYINT
    display_order = db.Column(db.Integer, default=0)
    created_at = db.Column(db.DateTime, default=func.now(), nullable=True)
    updated_at = db.Column(db.DateTime, default=func.now(), onupdate=func.now(), nullable=True)

    # Relationships
    product = db.relationship('Product', back_populates='images')
    
    def to_dict(self):
        """Convert product image object to dictionary"""
        return {
            'id': self.id,
            'product_id': self.product_id,
            'image_url': self.image_url,
            'is_primary': self.is_primary,
            'display_order': self.display_order,
            'created_at': self.created_at.isoformat() if self.created_at else None,
            'updated_at': self.updated_at.isoformat() if self.updated_at else None
        }
